client/cmd/dexc-desktop/main.go

```diff
@@ -0,0 +1,528 @@
+	if err != nil {
+		return fmt.Errorf("failed creating web server: %w", err)
+	}
+
+	webStart := make(chan error)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		cm := dex.NewConnectionMaster(webSrv)
+		webStart <- cm.Connect(appCtx)
+		cm.Wait()
+	}()
+
+	if err := <-webStart; err != nil {
+		return err
+	}
+
+	// No errors running webserver, so we can be certain we won any race between
+	// starting instances. Start the sync server now.
+	openC := make(chan struct{}) // no buffer. Is ignored if window is currently open
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		runServer(appCtx, syncDir, openC, killChan, cfg.Net)
+	}()
+
+	openWindow := func() {
+		wg.Add(1)
+		go func() {
+			runWebviewSubprocess(appCtx, "http://"+webSrv.Addr())
+			wg.Done()
+		}()
+	}
+
+	openWindow()
+
+	wg.Add(1)
+	go func() {
+		defer func() {
+			systray.SetTooltip("Shutting down. Please wait...")
+			systray.Quit()
+			wg.Done()
+		}()
+	windowloop:
+		for {
+			var backgroundNoteSent bool
+			select {
+			case <-windowManager.zeroLeft:
+			logout:
+				for {
+					err := clientCore.Logout()
+					if err == nil {
+						// Okay to quit.
+						break windowloop
+					}
+					if !errors.Is(err, core.ActiveOrdersLogoutErr) {
+						// Unknown error. Force shutdown.
+						log.Errorf("Core logout error: %v", err)
+						break windowloop
+					}
+					if !backgroundNoteSent {
+						sendDesktopNotification("DEX client still running", "DEX client is still resolving active DEX orders")
+						backgroundNoteSent = true
+					}
+					// Can't log out. Keep checking until either
+					//   1. We can log out. Exit the program.
+					//   2. The user reopens the window (via syncserver).
+					select {
+					case <-time.After(time.Minute):
+						// Try to log out again.
+						continue logout
+					case <-openC:
+						// re-open the window
+						openWindow()
+						continue windowloop
+					case <-appCtx.Done():
+						break windowloop
+					}
+				}
+			case <-appCtx.Done():
+				break windowloop
+			case <-openC:
+				openWindow()
+			}
+		}
+	}()
+
+	systray.Run(func() {
+		systrayOnReady(appCtx, filepath.Dir(cfg.LogPath), openC, killChan)
+	}, func() {
+		wg.Done()
+	})
+
+	closeAllWindows()
+
+	log.Infof("Shutting down...")
+	cancel()
+	wg.Wait()
+
+	return nil
+}
+
+var windowManager = &struct {
+	sync.Mutex
+	counter  uint32
+	windows  map[uint32]*exec.Cmd
+	zeroLeft chan struct{}
+}{
+	windows:  make(map[uint32]*exec.Cmd),
+	zeroLeft: make(chan struct{}, 1),
+}
+
+func closeWindow(windowID uint32) {
+	m := windowManager
+	m.Lock()
+	cmd, found := m.windows[windowID]
+	if !found {
+		m.Unlock()
+		// Probably killed by caller via closeAllWindows.
+		return
+	}
+	delete(m.windows, windowID)
+	remain := len(m.windows)
+	m.Unlock()
+	if remain == 0 {
+		select {
+		case m.zeroLeft <- struct{}{}:
+		default:
+		}
+	}
+	log.Infof("Closing window. %d windows remain open.", remain)
+	cmd.Process.Kill()
```

Comment:
If I shutdown before a webview is opened, I see a panic since `cmd.Process` will still be nill but have being added to `m.windows` giving other process an opportunity to access the`cmd.Process`.